From d043419fdd5662b59d891363de9755d171310c79 Mon Sep 17 00:00:00 2001
From: Alex_kl <iissmirnov1995@gmail.com>
Date: Sun, 22 Mar 2026 02:26:33 +0300
Subject: [PATCH 2/4] addmon was added, moving to monster position is now
 detected

---
 20260316/1/client.py | 90 ++++++++++++++++++++++++++++++++++++++++++--
 20260316/1/server.py | 15 ++++++++
 2 files changed, 102 insertions(+), 3 deletions(-)

diff --git a/20260316/1/client.py b/20260316/1/client.py
index f39749e..88f17b5 100644
--- a/20260316/1/client.py
+++ b/20260316/1/client.py
@@ -1,10 +1,24 @@
 import socket
 import sys
 import cmd
+import shlex
+import cowsay
 
 HOST = "localhost"
 PORT = 7778
 
+UNKNOWN_MONSTER = 1
+INVALID_ARGS = 2
+
+
+class Field_attributes:
+    def __init__(self):
+        self.monsters = cowsay.list_cows()
+        self.custom__monsters = {}
+        self.size_x = 10
+        self.size_y = 10
+        self.weapons = ["weapon", "axe", "sword"]
+
 class MUDCLient(cmd.Cmd):
     intro = "<<< Welcome to Python-MUD 0.1 >>>"
     prompt = ""
@@ -13,10 +27,10 @@ class MUDCLient(cmd.Cmd):
         self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
         self.sock.connect((host, port))
         self.sockfile = self.sock.makefile('r')
-        self._monster_names = None
-        self._custom_cows = {}
+        self.field = Field_attributes()
         super().__init__()
 
+
     def _send(self, msg):
         self.sock.sendall(msg.encode())
         response = self.sockfile.readline()
@@ -28,10 +42,71 @@ class MUDCLient(cmd.Cmd):
         self.sockfile.close()
         self.sock.close()
 
+
+
     def player_moving(self, dx, dy):
         cmd = "move " + str(dx) + " " + str(dy) + '\n'
         ans = self._send(cmd)
-        print(ans)
+        if '\t' in ans:
+            ans = ans.split('\t', maxsplit=2)
+            name = ans[1]
+            hello = ans[2]
+            if name in self.field.custom__monsters:
+                monster = self.field.custom__monsters[name]
+                print(cowsay.cowsay(hello, cowfile=monster))
+            else:
+                monster = name
+                print(cowsay.cowsay(hello, cow=monster))
+        else:
+            print(ans)
+
+    def field_addmon(self, line):
+        try:
+            line = shlex.split(line)
+        except ValueError:
+            raise ValueError(INVALID_ARGS)
+
+        if len(line) != 8:
+            raise ValueError(INVALID_ARGS)
+
+        if line[0] not in self.field.monsters and line[0] not in self.field.custom__monsters:
+            raise ValueError(UNKNOWN_MONSTER)
+        name = line[0]
+        
+        try:
+            ind = line.index('hello')
+            hello = line[ind + 1]
+            del line[ind:ind + 2]
+        except (IndexError, ValueError):
+            raise ValueError(INVALID_ARGS)
+
+        try:
+            ind = line.index('hp')
+            hp = line[ind + 1]
+            if hp.isdigit() and int(hp) > 0:
+                hp = int(hp)
+            else:
+                raise ValueError(INVALID_ARGS)
+        except (IndexError, ValueError):
+            raise ValueError(INVALID_ARGS)
+
+        try:
+            ind = line.index('coords')
+            x = int(line[ind + 1])
+            y = int(line[ind + 2])
+            if not(0 <= x < self.field.size_x and 0 <= y < self.field.size_y):
+                raise ValueError(INVALID_ARGS)
+        except (IndexError, ValueError):
+            raise ValueError(INVALID_ARGS)
+
+        cmd = f"addmon {name} {hp} {x} {y} {hello}\n"
+        ans = self._send(cmd)
+        if ans.startswith("Replaced the old monster"):
+            ans = ans.split('\t', maxsplit=1)
+            print(ans[1])
+            print(ans[0])
+        else:
+            print(ans)
 
     def do_down(self, arg):
         self.player_moving(0, 1)
@@ -45,6 +120,15 @@ class MUDCLient(cmd.Cmd):
     def do_right(self, arg):
         self.player_moving(1, 0)
 
+    def do_addmon(self, arg):
+        try:
+            self.field_addmon(arg)
+        except ValueError as error:
+            if error.args[0] == INVALID_ARGS:
+                print("Invalid arguments")
+            elif error.args[0] == UNKNOWN_MONSTER:
+                print("Cannot add unknown monster")
+
     def emptyline(self):
         pass
 
diff --git a/20260316/1/server.py b/20260316/1/server.py
index f332b79..6e445d3 100644
--- a/20260316/1/server.py
+++ b/20260316/1/server.py
@@ -43,6 +43,21 @@ class MUDServer:
             self.field.monster_types.append(data[1])
             msg = '0'
         
+        elif cmd == 'addmon':
+            name = data[1]
+            hp = int(data[2])
+            x, y = int(data[3]), int(data[4])
+            hello = ' '.join(data[5:])
+            stats = {
+                'name': name,
+                'hello': hello,
+                'hp': hp
+            }
+            fl_replaced = (x, y) in self.field.monsters
+            self.field.monsters[(x, y)] = stats
+            msg = f"Added monster {stats['name']} to ({x}, {y}) saying {stats['hello']}"
+            if fl_replaced:
+                msg = "Replaced the old monster" + '\t' + msg        
         msg += '\n'
         return msg
         
-- 
2.51.0

